firedrake/cofunction.py

```diff
@@ -79,6 +79,18 @@ def __del__(self):
         if hasattr(self, "_comm"):
             mpi.decref(self._comm)
 
+    def __call__(self, primal):
+        r"""Compute the duality pairing, this is to evaluate this
+        :class:`.Cofunction` at a :class:`.Function`, equivalent to taking the
+        dot product of the vectors of coefficients of the dual and primal
+        arguments.
```

Comment:
I think yhis is only true for euclidian vectors in R^n, not for Functions and Cofunctions in general. See my question about `ufl.action` below

Reply:
No that's wrong - you've got your cofunction using that inner product already